storage: space management config and health integration

dotnwat opened the pull request:
* First three commits are from PR https://github.com/redpanda-data/redpanda/pull/12132 under separate review which fix legacy_default config issue, but which this PR depends on.

* Then next four commits address integration with partition balancer by publishing size information into the cluster health report which the balancer will pick up. Used by @bharathv in separate PR.

* Final set of commits add / tweak configuration options to support (1) desired policies in upgrade scenarios and (2) percentage based expression of cache and log data disk size reservations.

Upgraded clusters
* The effect should be that new space management is not active
* In ordre to enable, turn the feature on and configure nonstrict-mode/disk sizes.

New clusters
* New space management is on by default.

Config option scenarios

|option|upgrade default|new-cluster default|
|----|---|---|
|local retention nonstrict|off|on|
|disk reservation|0|25%|
|log data capacity bytes|nullopt|nullopt|
|log data capacity pct|nullopt|80%|
|cache size bytes|10gb|0|
|cache size pct|nullopt|20%|
|feature toggle|off|on|


## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.1.x
- [ ] v22.3.x
- [ ] v22.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

dotnwat commented on tests/rptest/tests/archival_test.py:
```diff
@@ -205,7 +205,8 @@ def __init__(self, test_context):
         extra_rp_conf = dict(
             log_compaction_interval_ms=self.log_compaction_interval_ms,
-            log_segment_size=self.log_segment_size)
+            log_segment_size=self.log_segment_size,
+            retention_local_is_nonstrict=False)
```
This commit makes non-strict mode enabled by default for new clusters. This is causing problems for tests that have checks like "n segments deleted" or that truncation occurs up to some point. I started disabling nonstrict mode in some tests, but found that this category of check is all over. So it seems like it might be better to update the tests to deal with the new default, but it's not yet clear what those updates will look like.

VladLazar replied:
A lot of cloud storage ducktape tests use the truncation of segments as an indicator that those segments have been uploaded to the cloud. The correct thing do to here is to infer this from the manifest uploaded in the bucket/container. We have a utility for this `quiesce_uploads` (in `si_utils.py`) and some newer tests are already using it.

dotnwat replied:
DIscussed offline.

Currently approach is to disable the feature for the cloud tests which utilize this older way of inferring cloud usage by examining that truncation is occurring. We'll need to go back and update these tests individually in order to get expanded test coverage with the feature enabled. For now we'll then be relying on existing tests designed for the feature to exercise it.

VladLazar commented on src/v/config/configuration.cc:
```diff
@@ -1689,7 +1689,18 @@ configuration::configuration()
+      "disk space.",
+      {.needs_restart = needs_restart::no, .visibility = visibility::user},
+      20.0,
+      property<std::optional<double>>::noop_validator,
```
A validator that checks the [0, 100] bounds for these percent based configs would be great.

dotnwat replied:
fixed!

VladLazar commented on src/v/config/configuration.cc:
```diff
@@ -1689,7 +1689,18 @@ configuration::configuration()
+      0,
+      property<uint64_t>::noop_validator,
+      legacy_default<uint64_t>(20_GiB, legacy_version{9}))
+  , cloud_storage_cache_size_percent(
```
IIRC correctly, users can use a separate disk for the cloud storage cache. With that type of setup, this type of size allocation doesn't make sense. I don't think that's the case in our cloud, but I'd be curious if there's any self hosted customers that do this.

dotnwat replied:
Yes, that's correct. The percentage applies to either the shared data disk or the cache disk, but the default percent here is chosen assuming the shared disk setup.

We could do something automatic like switch it to 100% at runtime if a dedicated disk were chosen, but this seems like it might be confusing compared to solving this in documentation stating that this is the default in all cases, and if you use a dedicated disk, then you should consider increasing the value of this percentage tunable.

VladLazar commented on src/v/cluster/node/types.h:
```diff
@@ -56,12 +56,57 @@ struct local_state
+        uint64_t data_reclaimable_size{0};
+        friend std::ostream& operator<<(std::ostream&, const log_data_state&);
+    };
+    tristate<log_data_state> log_data_size{std::nullopt};
```
nit: why does this have to be a tristate? From the pov of the caller, the data is either available or not.

dotnwat replied:
Ahh, right. I will add a comment to explain this. We have three cases that we wanted to report to the balancer:

1. early bootup--health report produced, but space manager hasn't yet had the chance to figure out what's current state is.
2. configuration on the node has disabled the feature entirely
3. real results being reported

The distinction with (1) is that the balancer will wait until it gets more information before acting on state in order to avoid the case where we might report a large disk and usage information and then later when space manager starts changing that state is a drastic way.

dotnwat replied:
Update. After speaking with Bharath, we decided to go with this simpler interpretation of available or not available which also allows us to remove the feature flag. I think it's much simpler now to reason through. Thanks for point that out.

VladLazar commented on src/v/config/configuration.cc:
```diff
@@ -1628,7 +1630,22 @@ configuration::configuration()
+      {.needs_restart = needs_restart::no,
+       .example = "70.0",
+       .visibility = visibility::user},
+      80.0,
```
Shouldn't `retention_local_target_capacity_percent`, `disk_reservation_percent` and `cloud_storage_cache_size_percent` add up to 100 (for the case when the cloud storage cache uses the same disk)

Lazin replied:
I think we want to keep some free space. Flash translation layer of the SSD drive needs some space to be happy. Otherwise the latency may spike because of GC.

VladLazar replied:
Yup, I agree. Isn't `disk_reservation_percent` meant to be the slice of disk that we keep free though?

VladLazar reviewed (commented):
Generally looks good, but I have some questions

Lazin commented on src/v/cloud_storage/cache_service.cc:
```diff
@@ -62,26 +62,61 @@ static constexpr std::string_view tmp_extension{".part"};
   , _total_cleaned(0) {
     if (ss::this_shard_id() == ss::shard_id{0}) {
+        update_max_bytes(); // initialize _max_bytes
         _max_bytes_cfg.watch([this]() { update_max_bytes(); });
```
we should probably also watch `_max_percent` and `_disk_reservation`

dotnwat replied:
100% we should. thanks, I missed this.

andrwng commented on src/v/resource_mgmt/storage.cc:
```diff
@@ -110,13 +122,59 @@ ss::future<> disk_space_manager::run_loop() {
+
+    // bytes-based target size
+    const uint64_t target_size_bytes
+      = _retention_target_capacity_bytes().value_or(0);
```
Should this be capped to the usable size?

dotnwat replied:
fixed

andrwng commented on src/v/cloud_storage/cache_service.cc:
```diff
@@ -62,26 +62,61 @@ static constexpr std::string_view tmp_extension{".part"};
+    const uint64_t max_size_pct = usable_size
+                                  * (_max_percent().value_or(0) / 100.0);
+
+    const auto max_size_bytes = _max_bytes_cfg();
```
Same here w.r.t usable size?

dotnwat replied:
fixed

dotnwat commented on src/v/resource_mgmt/storage.cc:
```diff
@@ -95,6 +97,18 @@ ss::future<> disk_space_manager::run_loop() {
+        /*
+         * waiting for the nodes to upgrade isn't important locally, but it
+         * allows us to send a consistent message back to the partition balancer
+         * which may be trying to distinguish between new-blaancer / old-broker
```
fixed

mattschumpert commented on src/v/config/configuration.cc:
```diff
@@ -1618,7 +1618,9 @@ configuration::configuration()
       "inform data removal policies in low-disk space scenarios.",
       {.needs_restart = needs_restart::no, .visibility = visibility::user},
-      false)
+      true,
```
Fair point.  The default on upgrade is to keep it strict but it doesn't matter which way we go.

In the long run retention_local_strict = false might make more sense.

dotnwat replied:
great idea

dotnwat replied:
fixed

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1969,13 +1969,22 @@ log make_disk_backed_log(
+     */
+    auto local_retention_cfg = apply_base_overrides(input_cfg);
+    local_retention_cfg = override_retention_config(local_retention_cfg);
+    const auto local_retention_offset
+      = co_await maybe_adjusted_retention_offset(local_retention_cfg);
```
Good question. I'm not super familiar with these and just wanted to preserve as much of the existing semantics as possible. I'll take an action item to follow up on this and find out.

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1969,13 +1969,22 @@ log make_disk_backed_log(
         max_offset = retention_offset(cfg);
     }
 
+    /*
+     * offset calculation for local retention reclaimable is different than the
+     * retention above and takes into account local retention advisory flag.
+     */
+    auto local_retention_cfg = apply_base_overrides(input_cfg);
+    local_retention_cfg = override_retention_config(local_retention_cfg);
```
I think this is correct as-is, but please let me know if you disagree. To expand on this in a way that might help clarify what the intention is here:

Normal GC housekeeping will trim down to either local retention when retention_local_strict is true or down to the consumable retention when retention_local_strict is false (i.e. partitions are allowed to grow past local retention).

Normal GC housekeeping always runs with or without space management features enabled.

The context of this comment is that space management is inquiring about the segments which it might want to remove to free up space. The space management system knows that GC will already be trimming down the partition according the retention policy and the strict flag. So it is only concerned with space it can reclaim _beyond_ that point which is down to the local retention level, which is why we don't want to take into consideration the strict flag.

Of course if the strict flag is true, then we expect that the normal GC housekeeping will be aggressive, and space management would then only be making decisions that went even further past local retention level.

dotnwat commented:
Force push:

Feedback

* Clamped requests for specific byte size reservations to the usable size (@andrwng )
* Fixed missing binding watches (@Lazin )
* Changed name of advisory flag to `retention_local_strict` (@andrwng )
* Used optional instead of tristate to simplify state representation in health report (@VladLazar )
* Updated description of percentage based configurations (@andrwng @VladLazar @Lazin )
* Added bounds checking on configuration options (@VladLazar )
* Raised the control loop semaphore on all binding changes to give better responsive feedback when disabling space management.
* Applied the excellent advice to use SISettings to enable local retention strict mode in all cloud tests as a temporary workaround for test failures that use segment removal to infer data is uploaded. Also removed the previous per-test configuration changes that the changes to SISettings now handle (@VladLazar )

Other

* @bharathv and I decided that we could remove the the feature flag and simplify the interaction with the partition balancer, so we told that feature flag to get lost!

VladLazar commented on src/v/cluster/node/types.cc:
```diff
@@ -55,6 +68,11 @@ void local_state::serde_read(iobuf_parser& in, const serde::header& h) {
         }
     }
+
+    if (h._version >= 1) {
+        log_data_size = serde::read_nested<std::optional<log_data_state>>(
+          in, 0);
+    }
```
Since we've already branched out shouldn't we bump the serde version here?

dotnwat replied:
I fixed this

VladLazar commented on tests/rptest/services/redpanda.py:
```diff
@@ -564,6 +566,9 @@ def update_rp_conf(self, conf) -> dict[str, Any]:
             conf[
                 'cloud_storage_spillover_manifest_max_segments'] = self.cloud_storage_spillover_manifest_max_segments
+
+        conf['retention_local_strict'] = self.retention_local_strict
```
Pretty sure this overrides the `retention_local_strict` you are setting in `log_storage_target_size_test.py`. Effectively, both test variants will have `retention_local_strict=True. You should set `retention_local_strict` via `SISettings` or override it again after the call to `set_si_settings` in the test.

VladLazar reviewed (dismissed):
Looks good to me! Would be great to update the test in `log_storage_target_size_test.py` in order to have it run with `retention_local_strict=False` (see [this comment](https://github.com/redpanda-data/redpanda/pull/12091#discussion_r1266785798)).

dotnwat commented:
/backport v23.2.x

vbotbuildovich commented:
Failed to run cherry-pick command. I executed the commands below:

```
git checkout -b backport-pr-12091-v23.2.x-622 remotes/upstream/v23.2.x
git cherry-pick -x 820d15b44035d87bef02b574964a51c72a7144ec 743ba673dba625d79e03526c0dca29868b82a30c 862be40670b7da0cac4c3153da4e4c115c979b26 b7785e797eb5c83836966d2a481085db9ecd117a 3f307d186c23b13daff31f24a584448126994a3b 77d09e7f6fe55c069470cba6790d6ebce7f13b32 e186b7f41e3b6728b0e3ee580baf76e63a633c62 2065a7c0606a17afcad30c99c2a0394cf8fc9ae3 991daf1881cb708218061ae92052c21da9402bf4 773d35c2969082c06b45ba380067615352c839a2 61cc2fc41ed6c8797102a936a6b613d4f7e892f3 6c71b3d0728007bcb1df7da5604e34bcc759739f 6f70d0031ec63b232e693fbd5d8a71d13caebc54 bf81317c7964d7132be1ccc87d71b2a65eeccf7d 858647680cc6c1b7204f5368be7410c2b1494348 e90743a754f88b492b6803194978519a53e78958 dec333ac5fec67778421f4f2859c12503af72ff8 ba64314b964472fba7f17b9fa4a36e6594bc0c6b 39428697d259f807c1de5ed924f68cf3c787fd3e 5ba946425bd1c9556fdbca2f4d25e4eb0e5852a4 c099970a6d5762bf2880b8f8993f6d9c9c58641b 5b15c6bf3d56b98c6914badb3fa8c15837d1dc82 2f601644ba07fc3ecdcea983ec61366b19120cd4
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/5590297683).

dotnwat commented:
>Failed to run cherry-pick command. I executed the commands below:

dude

dotnwat commented:
https://github.com/redpanda-data/redpanda/pull/12286
